### ISSUE TITLE
BUG: `torch.cuda.is_available` is `False` with conda-forge install

### ISSUE THREAD

**Issue by NickleDave** [vocalpy/vak#470]
Installed with `conda create -n vak-tweetynet-tutorial vak tweetynet -c conda-forge`

Then did

```
$ conda activate vak-tweetynet-tutorial
(vak-tweetynet-tutorial) bart@tandy:~/Documents/repos/coding/birdsong/tweetynet-vak-tutorial
$ python
Python 3.9.10 | packaged by conda-forge | (main, Feb  1 2022, 21:24:11) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.cuda.is_available()
False
```

Tried installing `cudatoolkit` but this did not fix it:
```
$ conda install cudatoolkit=10.2 -c pytorch
Collecting package metadata (repodata.json): done
Solving environment: done
```

**Comment by NickleDave** [comment]
I tested this on an up-to-date pop-os and had the same issue.  
Should also test on Windows.

**Comment by NickleDave** [comment]
If I do `conda create -n vak-env vak -c conda-forge -c pytorch` on pop-os then `torch.cuda.is_available()` is now `True`

So maybe I just need to add that to install instructions?

Thought I tried that on Ubuntu though and it did *not* fix it

In general extra-verbose installs are annoying

**Comment by NickleDave** [comment]
Copying and pasting some helpful advice from @bollwyvl in the conda-forge gitter:


> hm... might just be how you're installing torch: I can get it to solve with something like:
>
> mamba create -n foo --dry-run vak 'pytorch=*=*cuda*'

> + pytorch                             1.10.2  cuda112py39h4de5995_1  conda-forge/linux-64      667MB
> + vak                                  0.4.1  pyhd8ed1ab_1           conda-forge/noarch         87kB


> as a pure python package, seems like it's slightly beneficial, as it would work on e.g. binder for demo purposes, even if it's no good for produciton

> you could, however, add a multi-output, and have yer own vak-gpu which further refined the dependency

> well, there's this: https://github.com/conda-forge/rift-feedstock/blob/main/recipe/meta.yaml

> there's a few ways to do a "pun" package (an output with the same name as the feedstock)

> yeah, looks like rift does that just fine

> so your vak-gpu might look like:

```
outputs:
- name: vak
- name: vak-gpu
  requirements:
    run:
      - {{ pin_subpackage("vak", max_pin="x.x") }}
      - cudnn
```

> or something

> but you probably won't be able to test it in CI


Thank you, I at least see that I can just add another name and separate run requirements
It's tricky because pytorch does this weird thing where you specify cudatoolkit as a separate package for a GPU install, or you specify cpuonly as a package (?) for CPU.

When installing from their channel.

Sorry if I'm telling you something you already know.

But it looks like rift-gpu is just, like rift with pytorch as the backend. I always need pytorch to be there :(

> right: you always want pytorch, but if you also require a "sentinel" cuda package, that might be enough

> nm, mamba create -n foo --dry-run vak pytorch cudnn does not blend

> so there's gotta be some other way, but I am not the guy to know

> well, there's pytorch-gpu, so mamba create -n foo --dry-run vak pytorch-gpu pytorch does the thing

> so your vak-gpu could depend on pytorch-gpu and vak

**Comment by NickleDave** [comment]
Looking at what other libraries do
* speechbrain: no conda install -- https://github.com/speechbrain/speechbrain#quick-installation
* pyannote: verbose conda install -- https://github.com/pyannote/pyannote-audio#installation
* fastai: has their own channel, but install looks terse -- https://github.com/fastai/fastai#installing

The `fastai` approach might be the least confusing for downstream users. So we'd need to set up a channel and write a recipe.  
I don't see a separate recipe repo or any GitHub Actions that build it?
https://github.com/fastai/fastai/blob/master/environment.yml

**Comment by bollwyvl** [comment]
To pick up: does the project have _any_ value (e.g. is its test suite functional) if installed next to a cpu version of pytorch?

you could consider inverting the output described above, using `run_constrained` as a negative solver clause:

```yaml
outputs:
- name: vak # default, tiny metapackage
  requirements:
    run: 
      - {{ pin_subpackage("vak-core", max_pin="x.x" }}
    run_constrained:
      - pytorch-cpu <0.0.0 # what? that's impossible!
- name: vak-core # the actual package
  requirements:
     run: 
       - pytorch
       # ... all the other stuff
```

This is similar to how `matplotlib` works: for historical reasons, it ships with the qt backend, but all downstream dependencies are encouraged to use `matplotlib-base`.

> set up a channel

Going off the conda-forge channel will eventually lead to a higher maintenance burden... as long as you can get away with doing pypi releases, the CF bots will do their best take care of your conda best practice needs.

**Comment by bollwyvl** [comment]
(though there's nothing wrong with encouraging the `pytorch` channel, as they have the bodies to throw at keeping stuff running)

**Comment by NickleDave** [comment]
Thank you @bollwyvl 

> To pick up: does the project have any value (e.g. is its test suite functional) if installed next to a cpu version of pytorch?

the main use case built in right now can be run on CPU, so it definitely has value with pytorch-cpu, which is great for people that may not have access to GPUs

> Going off the conda-forge channel will eventually lead to a higher maintenance burden... as long as you can get away with doing pypi releases, the CF bots will do their best take care of your conda best practice needs.

I hear you but watching people do installs I realize the main group I'm trying to reach will be just completely overwhelmed by any nuance like "add multiple channels" and "you have to do this on Windows but not on Mac".

Not sure if I can even automate away those nuances, either with CF or without.  
Need to think about it more but adding a `vak-gpu` might be the way

**Comment by NickleDave** [comment]
https://github.com/NickleDave/vak/issues/470#issuecomment-1070822941
> If I do conda create -n vak-env vak -c conda-forge -c pytorch on pop-os then torch.cuda.is_available() is now True

Tried this on ubuntu instead of `pop-os` but `conda create` runs forever at the "Solving environment" step.
Ubuntu 16.04, conda 4.8 and 4.11

**Comment by bollwyvl** [comment]
hm, the order of arguments _definitely_ matters, at least for channels, but might matter overall. 

not helping on the "simple invocation" front, but maybe try:

```bash
conda create --dry-run -n vak-env -c pytorch -c conda-forge -c nodefaults vak
                                   ^                         ^
                                   + prefer pytorch for GPU  + avoid ToS entanglements
```

and of course, `mamba` can _easily_ save minutes on big solves, especially on windows.

however... `conda/mamba/pip install` on top of existing environments is the devil... especially mixing them indiscriminately. you might want to encourage users to start heading down the annotated `environment.yml` route, e.g.

```yaml
# environment-vak.yml
channels:
- pytorch      # prefer pytorch for GPU builds, if available
- conda-forge  # use conda-forge for all others
- nodefaults   # avoid ToS-encumbered dependencies
dependencies:
- vak
# - your other conda dependencies
- pip
- pip:
  # - your pip dependencies, if you _need_ them 
```

and then...

```bash
conda env update --file environment-vak.yml
```

**Comment by NickleDave** [comment]
Thank you @bollwyvl 
> hm, the order of arguments _definitely_ matters, at least for channels, but might matter overall.
> 

I should know this by now

> not helping on the "simple invocation" front, but maybe try:
> 
> ```shell
> conda create --dry-run -n vak-env -c pytorch -c conda-forge -c nodefaults vak
>                                    ^                         ^
>                                    + prefer pytorch for GPU  + avoid ToS entanglements
> ```

by ToS you mean Anaconda's? Any reason in particular I should be worried about that?  
I guess leaving out `defaults` also makes the solve faster?

> and of course, `mamba` can _easily_ save minutes on big solves, especially on windows.
> 

Agreed. I should probably add a `:::note` suggesting to upgrade to 4.12 and use the new solver, at least.

> however... `conda/mamba/pip install` on top of existing environments is the devil... especially mixing them indiscriminately. you might want to encourage users to start heading down the annotated `environment.yml` route, e.g.
> 
> ```yaml
> # environment-vak.yml
> channels:
> - pytorch      # prefer pytorch for GPU builds, if available
> - conda-forge  # use conda-forge for all others
> - nodefaults   # avoid ToS-encumbered dependencies
> dependencies:
> - vak
> # - your other conda dependencies
> - pip
> - pip:
>   # - your pip dependencies, if you _need_ them 
> ```
> 
> and then...
> 
> ```shell
> conda env update --file environment-vak.yml
> ```

Thank you. Agreed some advice on envs helps but again most of our users so far are just learning that this concept even exists so I am afraid to overload  the docs page with this. Maybe a link to another page on envs would be good later

**Comment by NickleDave** [comment]
> hm, the order of arguments definitely matters, at least for channels, but might matter overall.

I tried switching with pytorch first but "Solving environment" still seemed to hang

Then I tried specifying a Python and that resolved. It also seems to have installed `cudatoolkit` this time?
```
conda create -n vak-env python=3.8 vak -c pytorch -c conda-forge
```

so:
- [ ] specify a python with `conda create` command/snippet in docs, as above

Note: we do have a lower bound on dependencies so that's not something making the solve slower

**Comment by NickleDave** [comment]
Yes, this works on Ubuntu 16.04.7 LTS:

```
conda create -n vak-env python=3.8 vak -c pytorch -c conda-forge
```

After `conda activate vak-env`:
```
(vak-env) art@CrabLab:~$ python
Python 3.8.12 | packaged by conda-forge | (default, Jan 30 2022, 23:42:07)
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.cuda.is_available()
True
```

- [ ] test with the channels reversed to be sure of that

**Comment by NickleDave** [comment]
Question is whether I could have a `vak` (implicit cpu) that installs `cpuonly` from `pytorch` channel, and then a `vak-gpu` that installs `cudatoolkit` etc., that would at least slightly reduce some of the verboseness?

**Comment by NickleDave** [comment]
closing for now -- I think for most users just adding `-c pytorch -c conda-forge` will get them what they want; will fix this in the docs in #527. If all else fails they can install `torch` + `torchvision` first exactly the way they need it, will document that too as a "troubleshooting" thing
